### DISCUSSION TITLE
building a custom model

### DISCUSSION THREAD

**Discussion by tve** [blakeblackshear/frigate#1426] [General]
Is anyone else building a custom model for frigate? I know about #1043, but that's not really what I'm looking for. I live in a rural setting and would like a model that can detect deer, horse, raccoon, skunk, squirrel, rat, mouse (the live ones), bobcat, mountain lion, coyote, etc. Also, I have a camera pointed at a bird bath and would love to get a list of species that visited each day. I don't expect that anything coming out of #1043 would help me...

I'm a noob at AI so I has taken me a lot of time to figure anything out, hence my posting here, 'cause it would be awesome to share experience. I have two coral edgetpu's, one for "production" and one for "dev & test" in a different system. So far I've been focusing on building a bird species object detector 'cause my bird bath has constant traffic, which would make it easy to see whether it actually works. Here are the steps I've taken:
- I'm starting with the [Caltech UCSB bird dataset](http://www.vision.caltech.edu/visipedia/CUB-200-2011.html) 'cause it has bounding boxes so lends itself to object detection (as opposed to classification)
- I'm using google colab for the training 'cause I don't have the kind of GPU resources necessary to get anywhere fast, I ended up subscribing to "colab pro" for $10/mo 'cause the free is too limited once I got to thousands of images
- I spent days trying various tutorials in vain, the whole tf1/tf2/tflite/coral x keras/model_maker/coco-formats jungle is the most confusing mess I've ever seen
- I have something working based on an [object detection tutorial by coral](https://colab.research.google.com/github/tensorflow/tensorflow/blob/master/tensorflow/lite/g3doc/tutorials/model_maker_object_detection.ipynb)
- That tutorial uses an EfficientDet-Lite model, which looks very promising, and it compiles to EdgeTPU, which is a requirement
- I've been trying to transfer-train the model using Cloud TPUs, but so far no luck, so I'm stuck with GPU (https://discuss.tensorflow.org/t/need-help-training-with-modelmaker-cloud-tpu-in-colab/3144/2)
- My first attempt at transfer training resulted in a rather poor mAP of 17%, I'm now trying to figure out whether I can use FiftyOne to visualize what I've got so I can perhaps see what's going on.

I'm hoping some of you are also somewhere on a similar journey and we can perhaps compare notes?

**Comment by trythatagain** [comment]
This is just what I've started to look at, we have a Hedgehog feeding station, and would like to an eye on it, we're in a battle with a fox who keeps nicking the food :)
So I'm going to follow this thread closely, as I'm more of a noobie than you, but I'm looking forward to it.
I've looked at a couple of youtube tutorials, but I'm just starting to get my head around it. I've looked at some picture recognition, where they have approx 1000 different  things that they can detect, which is 10 time we have with the object detection we are using in Frigate now.

**Reply by tve** [reply]
I found that the tutorials make it look deceptively simple...
One thing you will want to keep in mind is that object detection and image classification are quite different from each other. I don't know how the accuracy compares, but I suspect that a classification model trained with N images ends up being more accurate and faster than an object detection model trained with the same N images (plus bounding boxes). Currently Frigate requires an object detection model.

**Reply by trythatagain** [reply]
I agree with you about the tutorials. I think you're correct in the difference between image and object. There's a lot to take in :)

**Comment by tve** [comment]
I'm thinking more and more that it may make more sense to add an object classification stage to frigate after the object detection instead of trying to train a "supermodel". The model used by frigate needs to be retrained to detect some additional classes, such as deer, bobcat, rat, etc. But it's probably best to keep it relatively coarse-grained so the output remains something vague, such as person, rodent, bird, etc. This way a TFLite model running on a CoralTPU can be fast and still reasonably accurate.

Then those of us that want more fine-grained classification can run one or several additional models. The can either run on the CoralTPU or they could also run on the CPU using TF (non-lite) assuming the detections are infrequent enough (which they should be). Such additional models can, for example, classify trucks (Fed-Ex, UPS, ...), people (you, me, ...), license plates, birds, rodents, etc. They can be pure classification models, which should be more accurate for the same complexity than detection models, and for which there are often more/larger datasets.

So in the end, frigate would be a 3-stage system: motion, detection, classification. I still need more experimentation to figure out whether this can really work. One issue I see is that the "best" detection reported by the detector isn't always the best, so it may be necessary to extract a few more detection snapshots and classify each one and then pick the most likely.

**Reply by Honusnap** [reply]
Well .. we are in 2024 ... and i just installed Frigate and got the same feeling then you did in 2021 .... now the access to a better model cost 50$ ... and i can see where it is going, using users all over the world to work to build a model for a company that will pay big money for it.
I'm in the same situation as you were in 2021, i'm in a rural place and need to detect animals and people more then cars and fedex and those things. But the provided model .. and yolonas are really bad at this job.
Now i'm looking to give a shot at building my own model, did you built yours ?

**Comment by JeanDiv14** [comment]
I think I'm asking something similar.. just started using, tracking people, cats and birds and generating a counter in home assistant. Also a notification to text my neighbour when their cat is on my lawn :-) . We have a lot of squirrels, and I'd like to include a squirrel counter. I've found someone who has trained the model (here: https://github.com/rb2k/sqwrlmodel).. but how do I merge this into the existing .tflite model..?

Apologies if it's a silly question.

Thanks,
Jean

**Reply by tve** [reply]
Not a silly question, just no simple answer, alas... I'm on a trip and had to take a bit of a break from this but will be back on it in a couple of weeks...

**Reply by JeanDiv14** [reply]
Thank you!

**Reply by ozett** [reply]
multiple models seems only still a request ...
https://github.com/blakeblackshear/frigate/issues/1301

**Comment by ozett** [comment]
Feature Request:
i vote for a second AI stage. To try custom models and dont loose original functionality.
there should be an option to feed recognized categroes from the frist stage/mode, like car, person, cat, dog. ... to a second stage.
for AI with a custom model. it than does not hurt if there is anything wrong with the second stage. but its an addition win, if one can apply an seperate own model to the person, cat, dog... etc...

and train the second level like this:
https://github.com/blakeblackshear/frigate/discussions/1043#discussioncomment-1248161

**Comment by ozett** [comment]
https://www.tensorflow.org/lite/guide/model_maker   => 
👍  https://www.tensorflow.org/lite/tutorials/model_maker_image_classification

Model maker found in this thread:
https://blog.ambianic.ai/2021/09/02/movenet-vs-posenet-person-fall-detection.html
